state/protocol/badger/validity.go

```diff
@@ -199,29 +200,44 @@ func isValidEpochCommit(commit *flow.EpochCommit, setup *flow.EpochSetup) error
 	return nil
 }
 
+// isValidVersionBeacon validates internal structure of flow.VersionBeacon and in a context of given block,
+// which is assumed to be a block flow.VersionBeacon has been put into effect.
+// InvalidServiceEventError with appropriate message is returned if any validation fails.
+// Please refer to documentation of flow.VersionBeacon for a description of what's expected of a valid VersionBeacon.
 func isValidVersionBeacon(vb *flow.VersionBeacon, header *flow.Header) error {
 	if len(vb.RequiredVersions) == 0 {
 		return fmt.Errorf("required versions empty")
 	}
 
-	// first entry in a table must be a current version, so the height must be below the current block
-	if vb.RequiredVersions[0].Height > header.Height {
-		return fmt.Errorf("lowest required version height %d below current block's height %d", vb.RequiredVersions[0].Height, header.Height)
-	}
+	minimumThresholdHeight := header.Height + flow.VersionThreshold
 
 	// handle case when only one version is present
 	if len(vb.RequiredVersions) == 1 {
+		if vb.RequiredVersions[0].Height < minimumThresholdHeight {
+			return protocol.NewInvalidServiceEventError("required version (index=%d) height %d below minimum possible height (current block %d + threshold %d = %d)", 0, vb.RequiredVersions[0].Height, header.Height, flow.VersionThreshold, minimumThresholdHeight)
+		}
 		_, err := validateRequirement(vb.RequiredVersions[0])
 		return err
 	}
 
+	// if first requirement below current height, second one must obey threshold
+	if vb.RequiredVersions[0].Height <= header.Height {
+		if vb.RequiredVersions[1].Height < minimumThresholdHeight {
+			return protocol.NewInvalidServiceEventError("required version (index=%d) height %d below minimum possible height (current block %d + threshold %d = %d)", 1, vb.RequiredVersions[1].Height, header.Height, flow.VersionThreshold, minimumThresholdHeight)
+		}
```

Comment:
I'm having a hard time convincing myself that these conditions will always hold. I think they depend on safeguards implemented in the smart contract, to prevent the creation of subsequent version requirements within a certain number of blocks of one another. 

Here's an example where I think we could run into validation problems here:
- at block 100, add version requirement (height=2000, version=v2) via event `vb1`
- at block 200, add version requirement (height=2100, version=v3) via `vb2`
- at block 2050, add version requirement (height=4000, version=v4) via `vb3`

When we are processing `vb3`, it would contain this version table:
- Height=2000, Version=v2
- Height=2100, Version=v3
- Height=4000, Version=v4

Since the first requirement is below current height, we would trigger the condition on line 225, which would fail because the second version requirement's height is only 50 ahead of `head`. 

As mentioned in the previous comment, service event validation is a very sensitive codepath. Processing a service event which fails validation can cause the chain to halt, unless otherwise handled. Because of this, I think it would be worthwhile in general to:
- tend toward being more lenient on inputs here, so long as they are safe to process
- clearly document how the smart contract guarantees that all conditions for validation will always pass

Reply:
One problem with leniency is that if we miss the version or run a wrong one we risk permanent damage to a node (as in requiring manual intervention/maybe rollback).
In the situation you described update at `vb3` would be considered invalid, as we don't want modifications within 1000 blocks, which include adding versions further away. Or maybe its better described as freezing the table if a new version is expected to take place within 1000 blocks.
Otherwise, as you correctly noted, we would need to track when each requirement was added, and this would overly complicate things. 
Given projected usage - having a bit restricted window to make changes is a good tradeoff for simplified design.
I would however love to to hear your proposal what would be the best way to handle invalid events without crashing the chain.

Reply:
> One problem with leniency is that if we miss the version or run a wrong one we risk permanent damage to a node

Let's define what outcomes might happen here:
1. A version beacon may be discarded, resulting in failure to upgrade to the desired version
2. A version beacon may be incorrectly interpreted, resulting in an upgrade to an undesired version
3. A version beacon may be inconsistently interpreted by different nodes, resulting in different nodes executing the same chunks with different versions.

When you mention permanent damage, I think this is referring to Outcome 3. In that case a node may produce an Execution Result or Approval which is incorrect, creating evidence for them to be slashed (and in case of ENs polluting their execution state going forward). In Outcomes 1 and 2, all nodes agree on the version, even though the version might be different than specified in the VersionBeacon. Of course if there is a bug causing a VersionBeacon to be interpreted incorrectly (Outcome 2), we should fix this. By leniency, I mean handle more potential failures gracefully rather than throwing an error which will cause the chain to halt, not introducing bugs. From my perspective, Outcome 1 is relatively harmless and greatly preferred to halting the chain.

All nodes must process blocks in a consistent manner, and agree on the resulting protocol state after processing each block. We can say that each node executes a function `f(old_state,next_block)->new_state`. `f` must be deterministic, irrespective of the Version Beacon. Whether `f` is strict or lenient with respect to its inputs is unrelated to whether `f` is deterministic. So I don't think we need to worry about Outcome 3 here.

> I would however love to to hear your proposal what would be the best way to handle invalid events without crashing the chain.

I think a reasonable choice would be to discard all VersionBeacon events which fail validation (Outcome 1). Ideally, future VersionBeacon's could still be applied (the sequence number currently prevents this, I believe).